include/unifex/finally.hpp

```diff
@@ -0,0 +1,679 @@
+        auto* op = op_;
+
+        auto& completionOp = op->completionErrorOp_.template get<
+            operation_t<CompletionSender, finally_error_receiver>>();
+        completionOp.destruct();
+
+        // Discard existing stored error from source-sender.
+        auto& errorStorage = op->error_.template get<Error>();
+        errorStorage.destruct();
+
+        unifex::set_error(
+            static_cast<Receiver&&>(op->receiver_),
+            static_cast<OtherError&&>(otherError));
+      }
+
+      void set_done() && noexcept {
+        auto* op = op_;
+
+        auto& completionOp = op->completionErrorOp_.template get<
+            operation_t<CompletionSender, finally_error_receiver>>();
+        completionOp.destruct();
+
+        // Discard existing stored error from source-sender.
+        auto& errorStorage = op->error_.template get<Error>();
+        errorStorage.destruct();
+
+        unifex::set_done(static_cast<Receiver&&>(op->receiver_));
+      }
+
+      template <
+          typename CPO,
+          typename... Args,
+          std::enable_if_t<!is_receiver_cpo_v<CPO>, int> = 0>
+      friend auto tag_invoke(
+          CPO cpo,
+          const finally_error_receiver& r,
+          Args&&... args) noexcept(std::
+                                       is_nothrow_invocable_v<
+                                           CPO,
+                                           const Receiver&,
+                                           Args...>)
+          -> std::invoke_result_t<CPO, const Receiver&, Args...> {
+        return static_cast<CPO&&>(cpo)(
+            static_cast<const Receiver&>(r.op_->receiver_),
+            static_cast<Args&&>(args)...);
+      }
+
+      template <typename Func>
+      friend void tag_invoke(
+          tag_t<visit_continuations>,
+          const finally_error_receiver& r,
+          Func&& func) {
+        std::invoke(func, static_cast<const Receiver&>(r.op_->receiver_));
+      }
+
+      friend unstoppable_token tag_invoke(
+          tag_t<get_stop_token>, const finally_error_receiver&) noexcept {
+        return {};
+      }
+
+    private:
+      operation_type* op_;
+    };
+
+    template <
+        typename SourceSender,
+        typename CompletionSender,
+        typename Receiver>
+    class finally_done_receiver {
+      using operation_type =
+          finally_operation<SourceSender, CompletionSender, Receiver>;
+
+    public:
+      explicit finally_done_receiver(operation_type* op) noexcept : op_(op) {}
+
+      finally_done_receiver(finally_done_receiver&& other) noexcept
+        : op_(std::exchange(other.op_, nullptr)) {}
+
+      void set_value() && noexcept {
+        auto* op = op_;
+        op->completionDoneOp_.destruct();
+        unifex::set_done(static_cast<Receiver&&>(op->receiver_));
+      }
+
+      template <
+          typename Error,
+          std::enable_if_t<
+              std::is_invocable_v<decltype(unifex::set_error), Receiver, Error>,
```

Comment:
I've been thinking that a formulation of `is_tag_invocable_v` such as the following would be handy:

```c++
template<auto& Tag, class...As>
inline constexpr bool is_tag_invocable_v =
  std::is_invocable_v<decltype(Tag), As...>;
```

Would change the above to `is_tag_invocable_v<unifex::set_error, Receiver, Error>`.

Reply:
We already have `is_tag_invocable_v<CPO, Args...>` to be a query that checks whether you can call `tag_invoke(cpo, args...)`.

But perhaps this could be named `is_cpo_invocable_v`?